pypuf/learner/evolution_strategies/becker.py

```diff
@@ -0,0 +1,311 @@
+        combiner = LTFArray.combiner_xor
+        pop_size = np.shape(individuals)[0]
+        for i in range(pop_size):
+            yield LTFArray(np.matrix(individuals[i, :]), transform, combiner,
+                           bias=False)
+
+    @staticmethod
+    def get_delay_differences(ltf_arrays, pop_size, challenges, challenge_num):
+        # returns 2D array of delay differences for all challenges on every
+        #   individual
+        delay_diffs = np.empty((pop_size, challenge_num))
+        for i, ltf_array in enumerate(ltf_arrays):
+            for j, challenge in challenges:
+                delay_diffs[i][j] = ltf_array.val(challenge)
+        return delay_diffs
+
+    @staticmethod
+    def get_reliabilities(delay_diffs, epsilon_theo):
+        # returns 2D array of reliabilities for all challenges on every individual
+        reliabilities = np.zeros(np.shape(delay_diffs))
+        indices_of_unreliable = delay_diffs > -epsilon_theo and delay_diffs < epsilon_theo
+        reliabilities[indices_of_unreliable] = 1
+        return reliabilities
+
+    @staticmethod
+    def get_correlations(reliabilities, measured_rels):
+        # returns array of pearson correlation coefficients between reliability
+        #   array of individual and instance for all individuals
+        pop_size = np.shape(reliabilities)[0]
+        correlations = np.empty(pop_size)
+        for i in range(pop_size):
+            correlations[i] = np.corrcoef(reliabilities[i,:], measured_rels)[0, 1]
+        return correlations
+
+    @staticmethod
+    def sort_individuals(individuals, correlations):
+        # returns 2D array of individuals as given from input, but sorted through
+        #   correlation coefficients
+        sorted_indices = np.argsort(correlations)
+        return individuals[sorted_indices[::-1]]
+
+
+    # helping methods
+    @staticmethod
+    def learn_epsilon(instance, unreliability):
+        # returns epsilon on the assumption:
+        #   98% of challenges are reliable on a single PUF,
+        #   while there are 1/2 * ((96%)^k + 1) on k XORed PUFs
+        n = instance.n
+        k = instance.k
+        num = 1000
+        repeat = 100
+        quantile = ((1-2*unreliability)**k + 1) / 2
+        cut = num - int(num * quantile)
+        challenges = tools.sample_inputs(n, num)
+        ratios = np.zeros(num)
+        responses = np.zeros((num, repeat))
+        for i, challenge in enumerate(challenges):
+            for j in range(repeat):
+                responses[i, j] = instance.eval(challenge)
+            ratio = np.sum(responses[i, :]) / repeat
+            ratios[i] = ratio if ratio <= 0.5 else (1 - ratio)
+        ratios = np.sort(ratios)
+        return np.sum(ratios[cut-3 : cut+3]) / 10
+
+    @staticmethod
+    def get_epsilon_theo(n, unreliability):
+        return np.sqrt(n) * stats.norm.ppf(0.5 + unreliability/2)
+
+    @staticmethod
+    def set_pole_of_LTFs(instance, different_LTFs):
```

Comment:
As far as I understand, this method tries to set the sign of each single LTF correctly. However, we only need to decide if we need to negate the entire result or not, so that'd be a lot faster in terms of runtime.